# K-Nearest Neighbor Lab
Read over the sklearn info on [nearest neighbor learners](https://scikit-learn.org/stable/modules/neighbors.html#nearest-neighbors-classification)




In [6]:
from sklearn.neighbors import KNeighborsClassifier, KNeighborsRegressor
import numpy as np
import pandas as pd
from scipy.io import arff
import matplotlib.pyplot as plt
from sklearn.model_selection import train_test_split

## 1 K-Nearest Neighbor (KNN) algorithm

### 1.1 (15%) Basic KNN Classification

Learn the [Glass data set](https://archive.ics.uci.edu/dataset/42/glass+identification) using [KNeighborsClassifier](https://scikit-learn.org/stable/modules/generated/sklearn.neighbors.KNeighborsClassifier.html#sklearn.neighbors.KNeighborsClassifier) with default parameters.
- Randomly split your data into train/test.  Anytime we don't tell you specifics (such as what percentage is train vs test) choose your own reasonable values
- Give typical train and test set accuracies after running with different random splits
- Print the output probabilities for a test set (predict_proba)
- Try it with different p values (Minkowskian exponent) and discuss any differences

In [16]:
# Learn the glass data
from ucimlrepo import fetch_ucirepo 
from sklearn.neighbors import KNeighborsClassifier
  
# fetch dataset 
glass_identification = fetch_ucirepo(id=42) 
  
# data (as pandas dataframes) 
X = glass_identification.data.features 
y = glass_identification.data.targets 


trainTestSplits = [(0.2, 0.8), (0.3, 0.7), (0.4, 0.6)]

for split in trainTestSplits:

    trainingScores = []
    testScores = []
    for i in range(10):
        X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=split[0], train_size=split[1])

        classifier = KNeighborsClassifier()
        classifier.fit(X_train, y_train)

        testScores.append(classifier.score(X_test, y_test))
        trainingScores.append(classifier.score(X_train, y_train))

    print(f'average training score accuacy (train/test split {split}):', sum(trainingScores)/len(trainingScores))
    print(f'average test score accuracy (train/test split {split}):', sum(testScores)/len(testScores))


C:\Users\child\AppData\Roaming\Python\Python312\site-packages\sklearn\neighbors\_classification.py:238: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples,), for example using ravel().
  return self._fit(X, y)
C:\Users\child\AppData\Roaming\Python\Python312\site-packages\sklearn\neighbors\_classification.py:238: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples,), for example using ravel().
  return self._fit(X, y)
C:\Users\child\AppData\Roaming\Python\Python312\site-packages\sklearn\neighbors\_classification.py:238: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples,), for example using ravel().
  return self._fit(X, y)
C:\Users\child\AppData\Roaming\Python\Python312\site-packages\sklearn\neighbors\_classification.py:238: DataConversionWarning: A column-vector y was

average training score accuacy (train/test split (0.2, 0.8)): 0.7631578947368421
average test score accuracy (train/test split (0.2, 0.8)): 0.6534883720930232


C:\Users\child\AppData\Roaming\Python\Python312\site-packages\sklearn\neighbors\_classification.py:238: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples,), for example using ravel().
  return self._fit(X, y)
C:\Users\child\AppData\Roaming\Python\Python312\site-packages\sklearn\neighbors\_classification.py:238: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples,), for example using ravel().
  return self._fit(X, y)
C:\Users\child\AppData\Roaming\Python\Python312\site-packages\sklearn\neighbors\_classification.py:238: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples,), for example using ravel().
  return self._fit(X, y)
C:\Users\child\AppData\Roaming\Python\Python312\site-packages\sklearn\neighbors\_classification.py:238: DataConversionWarning: A column-vector y was

average training score accuacy (train/test split (0.3, 0.7)): 0.7657718120805369
average test score accuracy (train/test split (0.3, 0.7)): 0.6415384615384616


C:\Users\child\AppData\Roaming\Python\Python312\site-packages\sklearn\neighbors\_classification.py:238: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples,), for example using ravel().
  return self._fit(X, y)
C:\Users\child\AppData\Roaming\Python\Python312\site-packages\sklearn\neighbors\_classification.py:238: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples,), for example using ravel().
  return self._fit(X, y)
C:\Users\child\AppData\Roaming\Python\Python312\site-packages\sklearn\neighbors\_classification.py:238: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples,), for example using ravel().
  return self._fit(X, y)
C:\Users\child\AppData\Roaming\Python\Python312\site-packages\sklearn\neighbors\_classification.py:238: DataConversionWarning: A column-vector y was

average training score accuacy (train/test split (0.4, 0.6)): 0.75
average test score accuracy (train/test split (0.4, 0.6)): 0.6267441860465116


C:\Users\child\AppData\Roaming\Python\Python312\site-packages\sklearn\neighbors\_classification.py:238: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples,), for example using ravel().
  return self._fit(X, y)
C:\Users\child\AppData\Roaming\Python\Python312\site-packages\sklearn\neighbors\_classification.py:238: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples,), for example using ravel().
  return self._fit(X, y)


In [17]:
# Print the output probabilities for a test set (predict_proba)

# Learn the glass data
from ucimlrepo import fetch_ucirepo 
from sklearn.neighbors import KNeighborsClassifier
  
# fetch dataset 
glass_identification = fetch_ucirepo(id=42) 
  
# data (as pandas dataframes) 
X = glass_identification.data.features 
y = glass_identification.data.targets 

X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=split[0], train_size=split[1])


classifier = KNeighborsClassifier()
classifier.fit(X_train, y_train)

print('test set accuracy: ', classifier.score(X_test, y_test))
print('training set accuracy: ', classifier.score(X_train, y_train)) 
print('test set output probabilities: ', classifier.predict_proba(X_test))


test set accuracy:  0.686046511627907
training set accuracy:  0.6640625
test set output probabilities:  [[1.  0.  0.  0.  0.  0. ]
 [0.8 0.2 0.  0.  0.  0. ]
 [0.4 0.2 0.4 0.  0.  0. ]
 [0.8 0.  0.2 0.  0.  0. ]
 [0.  1.  0.  0.  0.  0. ]
 [0.4 0.4 0.2 0.  0.  0. ]
 [0.  0.8 0.  0.2 0.  0. ]
 [0.4 0.  0.6 0.  0.  0. ]
 [1.  0.  0.  0.  0.  0. ]
 [0.4 0.2 0.4 0.  0.  0. ]
 [0.  1.  0.  0.  0.  0. ]
 [1.  0.  0.  0.  0.  0. ]
 [0.  0.  0.  0.  0.  1. ]
 [0.  0.  0.  1.  0.  0. ]
 [0.2 0.2 0.4 0.  0.  0.2]
 [1.  0.  0.  0.  0.  0. ]
 [0.6 0.2 0.2 0.  0.  0. ]
 [0.6 0.2 0.2 0.  0.  0. ]
 [0.2 0.4 0.  0.  0.4 0. ]
 [0.6 0.4 0.  0.  0.  0. ]
 [0.2 0.4 0.4 0.  0.  0. ]
 [0.2 0.8 0.  0.  0.  0. ]
 [0.2 0.6 0.  0.  0.  0.2]
 [0.  0.  0.  0.  0.  1. ]
 [0.2 0.2 0.2 0.  0.2 0.2]
 [1.  0.  0.  0.  0.  0. ]
 [1.  0.  0.  0.  0.  0. ]
 [0.  0.8 0.  0.2 0.  0. ]
 [0.4 0.6 0.  0.  0.  0. ]
 [0.  0.  0.  0.  0.  1. ]
 [0.6 0.4 0.  0.  0.  0. ]
 [0.2 0.8 0.  0.  0.  0. ]
 [0.2 0.4 0.2 0.  0.2 0. ]
 [0.4

C:\Users\child\AppData\Roaming\Python\Python312\site-packages\sklearn\neighbors\_classification.py:238: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples,), for example using ravel().
  return self._fit(X, y)


In [19]:
# Try it with different p values (Minkowskian exponent) and discuss any differences

# Learn the glass data
from ucimlrepo import fetch_ucirepo 
from sklearn.neighbors import KNeighborsClassifier
  
# fetch dataset 
glass_identification = fetch_ucirepo(id=42) 
  
# data (as pandas dataframes) 
X = glass_identification.data.features 
y = glass_identification.data.targets 

X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=split[0], train_size=split[1])

pValues = [2, 3, 4, 5, 6, 8, 10]
for pValue in pValues:

    classifier = KNeighborsClassifier(p=pValue)
    classifier.fit(X_train, y_train)

    print(f'test set accuracy (p value: {pValue}): ', classifier.score(X_test, y_test))
    print(f'training set accuracy (p value: {pValue}): ', classifier.score(X_train, y_train)) 


C:\Users\child\AppData\Roaming\Python\Python312\site-packages\sklearn\neighbors\_classification.py:238: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples,), for example using ravel().
  return self._fit(X, y)
C:\Users\child\AppData\Roaming\Python\Python312\site-packages\sklearn\neighbors\_classification.py:238: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples,), for example using ravel().
  return self._fit(X, y)
C:\Users\child\AppData\Roaming\Python\Python312\site-packages\sklearn\neighbors\_classification.py:238: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples,), for example using ravel().
  return self._fit(X, y)


test set accuracy (p value: 2):  0.627906976744186
training set accuracy (p value: 2):  0.75
test set accuracy (p value: 3):  0.627906976744186
training set accuracy (p value: 3):  0.7265625
test set accuracy (p value: 4):  0.6744186046511628
training set accuracy (p value: 4):  0.734375
test set accuracy (p value: 5):  0.6744186046511628
training set accuracy (p value: 5):  0.7265625
test set accuracy (p value: 6):  0.6627906976744186
training set accuracy (p value: 6):  0.71875


C:\Users\child\AppData\Roaming\Python\Python312\site-packages\sklearn\neighbors\_classification.py:238: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples,), for example using ravel().
  return self._fit(X, y)
C:\Users\child\AppData\Roaming\Python\Python312\site-packages\sklearn\neighbors\_classification.py:238: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples,), for example using ravel().
  return self._fit(X, y)
C:\Users\child\AppData\Roaming\Python\Python312\site-packages\sklearn\neighbors\_classification.py:238: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples,), for example using ravel().
  return self._fit(X, y)


test set accuracy (p value: 8):  0.6627906976744186
training set accuracy (p value: 8):  0.7109375
test set accuracy (p value: 10):  0.6511627906976745
training set accuracy (p value: 10):  0.7109375


C:\Users\child\AppData\Roaming\Python\Python312\site-packages\sklearn\neighbors\_classification.py:238: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples,), for example using ravel().
  return self._fit(X, y)


**Discussion:**  
| Train/Test Split | Training Accuracy | Test accuracy |
| --- | --- | --- |
| (0.2, 0.8) | 0.7631578947368421 | 0.6534883720930232 |
| (0.3, 0.7) | 0.7657718120805369 | 0.6415384615384616 |
| (0.4, 0.6) | 0.75 | 0.6267441860465116 | 

As you can see above, I started by running the KNeighborsClassifier model on differen train/test splits. So, I ran with the splits (0.2, 0.8), (0.3, 0.7), and (0.4, 0.6). Even with these considerably different training/set sets, the accuracies were still very close to each other. All things considered, Training set accuracy was always about 76% and the test set accuracy was always about 64%. Finally, as expeced, the Training accuracy was always slightly higher than the Test accuracy because the training set is the group that the model was trained on, so it had seen it all before.  

After experimenting with different training/test splits, I output the probabilities for a test set (with predict_proba). Here are what some of those outputs looked like:  
[[0.8 0.2 0.  0.  0.  0. ]  
 [0.4 0.2 0.4 0.  0.  0. ]  
 [0.8 0.  0.2 0.  0.  0. ]  
 [0.  1.  0.  0.  0.  0. ]  
 [0.4 0.4 0.2 0.  0.  0. ]  
 [0.  0.8 0.  0.2 0.  0. ]  
 [0.4 0.  0.6 0.  0.  0. ]  
 [1.  0.  0.  0.  0.  0. ]  
 [0.4 0.2 0.4 0.  0.  0. ]  
 [0.  1.  0.  0.  0.  0. ]  
 [1.  0.  0.  0.  0.  0. ]  
 [0.  0.  0.  0.  0.  1. ]  
 [0.  0.  0.  1.  0.  0. ]  
 [0.2 0.2 0.4 0.  0.  0.2]  
 [1.  0.  0.  0.  0.  0. ]  
 [0.6 0.2 0.2 0.  0.  0. ]  
 [0.6 0.2 0.2 0.  0.  0. ]  
 [0.2 0.4 0.  0.  0.4 0. ]  
 [0.6 0.4 0.  0.  0.  0. ]  
 [0.2 0.4 0.4 0.  0.  0. ]  
 [0.2 0.8 0.  0.  0.  0. ]  
 [0.2 0.6 0.  0.  0.  0.2]  
...  
 [0.8 0.  0.2 0.  0.  0. ]  
 [0.8 0.2 0.  0.  0.  0. ]  
 [0.  0.6 0.  0.4 0.  0. ]  
 [0.6 0.2 0.2 0.  0.  0. ]]    

This matrix of values tells us for each instance, the likelyhood of that instance being classified in each output class. So, for example, the row [0.4 0.4 0.2 0.  0.  0. ] means that this instance had a 40% similarity to class 1, 40% similarity to class 2, 20% similarity to class 3, and a negligible similarity to the remaining output classes.  


Finally, I experimented with different p values (Minkowskian exponent):  
| p value | Training Accuracy | Test accuracy |
| --- | --- | --- |
| 2 | 0.75 | 0.627906976744186 |
| 3 | 0.7265625 | 0.627906976744186 |
| 4 | 0.734375 | 0.6744186046511628 |
| 5 | 0.7265625 | 0.6744186046511628 | 
| 6 | 0.71875 | 0.6627906976744186 | 
| 8 | 0.7109375 | 0.6627906976744186 | 
| 10 | 0.7109375 | 0.6511627906976745 |  

As you can see from the table above, as the p value increased, there was a slight decrease in accuracy for both Training and Test sets. The p value is the Minkowskian exponent, meaning that it factors into the Minkowskian equation and effects the distance calculated betwen points. So it looks like, at least for this dataset, the higher the exponent, the lower the accuracy. Perhaps this is because the more complicated distance equation is unessesary for such a simple dataset.


## 2 KNN Classification with normalization and distance weighting

Use the [magic telescope](https://axon.cs.byu.edu/data/uci_class/MagicTelescope.arff) dataset

### 2.1 (5%) - Without Normalization or Distance Weighting
- Do random 80/20 train/test splits each time
- Run with k=3 and *without* distance weighting and *without* normalization
- Show train and test set accuracy

In [1]:
# Learn magic telescope data

*Discussion*

### 2.2 (10%) With Normalization
- Try it with k=3 without distance weighting but *with* normalization of input features.  You may use any reasonable normalization approach (e.g. standard min-max normalization between 0-1, z-transform, etc.)

In [ ]:
# Train/Predict with normalization

*Discuss the results of using normalized data vs. unnormalized data*

### 2.3 (10%) With Distance Weighting
- Try it with k=3 and with distance weighting *and* normalization

In [ ]:
#Train/Precdict with normalization and distance weighting

Comparison and Discussion

### 2.4 (10%) Different k Values
- Using your normalized data with distance weighting, create one graph with classification accuracy on the test set on the y-axis and k values on the x-axis.
- Use values of k from 1 to 15.  Use the same train/test split for each. 

In [3]:
# Calculate and Graph classification accuracy vs k values

*Discussion*

## 3 KNN Regression with normalization and distance weighting

Use the [sklean KNeighborsRegressor](https://scikit-learn.org/stable/modules/generated/sklearn.neighbors.KNeighborsRegressor.html#sklearn.neighbors.KNeighborsRegressor) on the [housing price prediction](https://axon.cs.byu.edu/data/uci_regression/housing.arff) problem.  
### 3.1 (5%) Ethical Data
Note this data set has an example of an inappropriate input feature which we discussed.  State which feature is inappropriate and discuss why.

*Discuss the innapropriate feature*

### 3.2 (15%) - KNN Regression 
- Do random 80/20 train/test splits each time
- Run with k=3
- Print the score (coefficient of determination) and Mean Absolute Error (MAE) for the train and test set for the cases of
  - No input normalization and no distance weighting
  - Normalization and no distance weighting
  - Normalization and distance weighting
- Normalize inputs features where needed but do not normalize the output

In [6]:
# Learn and experiment with housing price prediction data

*Discuss your results*

### 3.3 (10%)  Different k Values
- Using housing with normalized data and distance weighting, create one graph with MAE on the test set on the y-axis and k values on the x-axis
- Use values of k from 1 to 15.  Use the same train/test split for each. 

In [ ]:
# Learn and graph for different k values

Discussion

## 4. (20%) KNN with nominal and real data

- Use the [lymph dataset](https://axon.cs.byu.edu/data/uci_class/lymph.arff)
- Use a 80/20 split of the data for the training/test set
- This dataset has both continuous and nominal attributes 
- Implement a distance metric which uses Euclidean distance for continuous features and 0/1 distance for nominal. Hints:
    - Write your own distance function (e.g. mydist) and use clf = KNeighborsClassifier(metric=mydist)
    - Change the nominal features in the data set to integer values since KNeighborsClassifier expects numeric features. I used Label_Encoder on the nominal features.
    - Keep a list of which features are nominal which mydist can use to decide which distance measure to use
    - There was an occasional bug in SK version 1.3.0 ("Flags object has no attribute 'c_contiguous'") that went away when I upgraded to the lastest SK version 1.3.1 
- Use your own choice for k and other parameters

In [5]:
# Train/Predict lymph with your own distance metric

*Explain your distance metric and discuss your results*

## 5. (Optional 15% extra credit) Code up your own KNN Learner 
Below is a scaffold you could use if you want. Requirements for this task:
- Your model should support the methods shown in the example scaffold below
- Use Euclidean distance to decide closest neighbors
- Implement both the classification and regression versions
- Include optional distance weighting for both algorithms
- Run your algorithm on the magic telescope and housing data sets above and discuss and compare your results 

*Discussion*

In [ ]:
from sklearn.base import BaseEstimator, ClassifierMixin

class KNNClassifier(BaseEstimator,ClassifierMixin):
    def __init__(self, columntype=[], weight_type='inverse_distance'): ## add parameters here
        """
        Args:
            columntype for each column tells you if continues[real] or if nominal[categoritcal].
            weight_type: inverse_distance voting or if non distance weighting. Options = ["no_weight","inverse_distance"]
        """
        self.columntype = columntype #Note This won't be needed until part 5
        self.weight_type = weight_type

    def fit(self, data, labels):
        """ Fit the data; run the algorithm (for this lab really just saves the data :D)
        Args:
            X (array-like): A 2D numpy array with the training data, excluding targets
            y (array-like): A 2D numpy array with the training targets
        Returns:
            self: this allows this to be chained, e.g. model.fit(X,y).predict(X_test)
        """
        return self
    
    def predict(self, data):
        """ Predict all classes for a dataset X
        Args:
            X (array-like): A 2D numpy array with the training data, excluding targets
        Returns:
            array, shape (n_samples,)
                Predicted target values per element in X.
        """
        pass

    #Returns the Mean score given input data and labels
    def score(self, X, y):
        """ Return accuracy of model on a given dataset. Must implement own score function.
        Args:
            X (array-like): A 2D numpy array with data, excluding targets
            y (array-like): A 2D numpy array with targets
        Returns:
            score : float
                Mean accuracy of self.predict(X) wrt. y.
        """
        return 0